Hw5

de-brasi opened the pull request.

de-brasi committed:
make url unique for table "links"

de-brasi committed:
Fix datatype for Telegram chat id

de-brasi committed:
move package "entities" to package "clients"

de-brasi committed:
using new location of UpdateResponse class (clients.entities instead of entities)

de-brasi committed:
better naming of package - move from "client" to common package "clientS"

de-brasi committed:
fix datatype for tg chat ids

de-brasi committed:
init entities

de-brasi committed:
init

de-brasi committed:
init;
todo: checkout "broken" error handling in JdbcLinkRepository::add and JdbcTelegramChatRepository::add (DataAccessException exception, raised after CAUGHT DataAccessException and printing marking-message)

de-brasi committed:
init

de-brasi committed:
undo excluding "org.springframework.boot.autoconfigure.jdbc.DataSourceAutoConfiguration" autoconfiguration from application.yml spring.autoconfigure.exclude - need only when DataSourceAutoConfiguration is not configured for project

de-brasi committed:
move from "domain" package to core module's package

de-brasi committed:
update path to entities

de-brasi committed:
init; structure source is template in task's description

de-brasi committed:
init; realization required

de-brasi committed:
make JdbcTemplate autowired in constructor

de-brasi committed:
add datasource configuration (url, username, password, driver-class-name);
delete excluding "org.springframework.boot.autoconfigure.jdbc.DataSourceAutoConfiguration" in autoconfigure

de-brasi committed:
add logger

de-brasi committed:
fix datasource url

de-brasi committed:
realization

de-brasi committed:
add logging exception and stacktrace for internalErrorHandler

de-brasi committed:
autowire TgChatService to constructor;
using TgChatService in handlers

de-brasi committed:
remove unnecessary autoconfiguration settings for jdbc datasource

de-brasi committed:
remove spring-boot-starter-jdbc and postgresql dependencies

de-brasi committed:
add spring-boot-starter-jdbc and postgresql dependencies

de-brasi committed:
throw IncorrectRequestException when MalformedURLException caught

de-brasi committed:
remove unnecessary todos

de-brasi committed:
delete @Nullable annotation from add method

de-brasi committed:
init

de-brasi committed:
init; need to fix code (in controller and service) and test "removeInCorrectPairLinkNotExists" - more info in test's todo

de-brasi committed:
init; draft version

de-brasi committed:
init

de-brasi committed:
better naming

de-brasi committed:
better naming for LinksNotExistException (to LinksNotAddedException)

de-brasi committed:
init;
add usage hints in todo!

de-brasi committed:
add usage hints in todo!

de-brasi committed:
init

de-brasi committed:
i think it's done, but not tested yet

de-brasi committed:
done

de-brasi committed:
better; add bugfixes; tested version

de-brasi committed:
add constructor from exception

de-brasi committed:
add methods realization

de-brasi committed:
better sql for getting all chat-link bounds

de-brasi committed:
handleGetLinks realization;
tested via swagger;
todo: need to get records only for specified id

de-brasi committed:
handlePostLinks realization;
tested via swagger;
todo: need to support correct link's id

de-brasi committed:
better naming

de-brasi committed:
discard stub entity

de-brasi committed:
untrackLinkForChat realization

de-brasi committed:
add lombok NoArgsConstructor

de-brasi committed:
TEMPORARY(!) comment test need to

de-brasi committed:
logging via log4j.Logger

de-brasi committed:
move from root to "clients" package

de-brasi committed:
using custom exceptions from package moved to

de-brasi committed:
move to package "domain"

de-brasi committed:
using entities ChatLinkBound, TelegramChat, Link from package "domain"

de-brasi committed:
rename field time-in-milliseconds to timeout-in-milliseconds (in properties file, config and beans)

de-brasi committed:
add base url getter

de-brasi committed:
make getters non static

de-brasi committed:
add fetchUpdate methods from url

de-brasi committed:
make botClient bean

de-brasi committed:
better api

de-brasi committed:
better (cascade deletion, unique pairs in track_info and etc)

de-brasi committed:
add fields for storing last checkout time and last value of "update" (got from api)

de-brasi committed:
change sql for inserting new url (add values for field last_check_time and last_update_time)

de-brasi committed:
using linkRepository for adding urls to links table

de-brasi committed:
realization; need to be tested

de-brasi committed:
better api

de-brasi committed:
add realization; need to be tested

de-brasi committed:
remove updating to LinkUpdater::update

de-brasi committed:
add realization for update method (iterate throw links in argument, check if updated, notify link's subscribers)

de-brasi committed:
add function to check is link supported by service

de-brasi committed:
remove unnecessary base url getter

de-brasi committed:
fix error messaging in retrieveLastActivityDateField

de-brasi committed:
better method "compareAndSetLastUpdateTime" - separate sql query: checkout last_update_time and if it outdated than update;
add method "checkLastUpdateTimeChanged";
add little dockstring to "notifyClientsIfUpdatedTimeChanged" and add calling dates updating methods ("actualizeLastUpdateTimeForLink" and "actualizeCheckingTimeForLink") only after success notifying subscribers via bot client;
add methods "actualizeLastUpdateTimeForLink" and "actualizeCheckingTimeForLink";

de-brasi committed:
delete unnecessary updating in "checkLastUpdateTimeChanged" method

de-brasi committed:
better interface; delete methods which should not be public

de-brasi committed:
using new LinkUpdater interface; delegate fetching links need to recheck to linkUpdater

de-brasi committed:
sync with new LinkUpdater interface

de-brasi committed:
linter fix

de-brasi committed:
linter fix

de-brasi committed:
discard unnecessary todos

de-brasi committed:
linter fix

de-brasi committed:
delete irrelevant tests (because they are tests interaction without database)

de-brasi committed:
change default value of timestamp on field 'last_update_time' when create new record in table 'last_update_time'

de-brasi committed:
delete test property source

de-brasi committed:
try to fix problem with not catching DataAccessException;
before creating new record count record with equal url;
add todo described problem with DataAccessException

de-brasi committed:
delete extra import

de-brasi committed:
try to fix problem with not catching DataAccessException;
before creating new record count record with equal chat_id;
add todo described problem with DataAccessException

de-brasi committed:
linter fix

de-brasi committed:
add method for searching by predicate for task 4

de-brasi committed:
add "search" method for task 4 in hw

de-brasi committed:
add "search" method for task 4 in hw; mark all public methods as transactional

de-brasi committed:
add "search" method for task 4 in hw;

de-brasi committed:
add @Override to "search" method

de-brasi committed:
using new methods JdbcLinkRepository::search for getting outdated links

de-brasi committed:
linter fix

de-brasi committed:
catching errors when getting links for update

de-brasi committed:
move try {...} catch() {...} from called JdbcLinkUpdater to caller LinkUpdaterScheduler;
main update script in JdbcLinkUpdater::update now called under try-catch

edikgoose commented on scrapper/src/main/java/edu/java/LinkUpdaterScheduler.java:
```diff
@@ -1,15 +1,54 @@
+        }
     }
+
+    private final static Logger LOGGER = LogManager.getLogger();
```
Почитай про @sl4j аннотацию. Удобная штука

de-brasi replied:
Ага, и правда очень удобно. Поправил

edikgoose commented on scrapper/src/main/java/edu/java/api/ChatController.java:
```diff
@@ -16,11 +18,19 @@
 public class ChatController {
+    private final TgChatService tgChatService;
+
+    public ChatController(@Autowired TgChatService tgChatService) {
```
Можно использовать RequiredArgsConstuctor анноташку, чтобы не писать конструктор. Позволяет не громодить код

de-brasi replied:
Спасибо за совет, поправил

edikgoose commented on scrapper/src/main/java/edu/java/api/ChatController.java:
```diff
@@ -30,7 +40,8 @@ public ResponseEntity<ApiErrorResponse> handleDeleteChat(@PathVariable Long id)
         // todo проверять на:
         //  - некорректные параметры 400
         //  - чат не существует 404
```
А ты это уже не сделал?

de-brasi replied:
Ага, да позабыл убрать тудушки после того как сделал)

edikgoose commented on scrapper/src/main/java/edu/java/api/LinksController.java:
```diff
@@ -19,45 +29,67 @@
 @RequestMapping(value = "/scrapper/api/links", produces = MediaType.APPLICATION_JSON_VALUE)
 @SuppressWarnings({"MultipleStringLiterals"})
 public class LinksController {
```
Классы, которые работают над сущностью, обычно используют ед число (LinkController)

de-brasi replied:
Поправил

edikgoose commented on scrapper/src/main/java/edu/java/api/LinksController.java:
```diff
@@ -19,45 +29,67 @@
 
-        return new ResponseEntity<>(LIST_LINKS_RESPONSE_STUB, HttpStatus.OK);
+        Collection<Link> allLinks = linkService.listAll(tgChatId);
+        // todo: добавить id в сущность Link, брать id оттуда
```
А в чем была проблема сразу его добавить? Очень непонятный момент, скушаю балл за это

de-brasi replied:
Данные о ссылке приходят к клиенту scrapper в таком порядке: БД -> Link(dto) -> Repository -> Services -> Controller -> обернуть в ListLinksResponse -> отправка клиенту по http. 
Id встречается в LinkResponse, который входит в ListLinksResponse. 
Когда делал дз3, было не понятно значение id чего (==откуда взятое) нужно передавать клиенту сервиса вместе с LinkResponse. Если этот id'шник - это id из базы данных, то сейчас тоже не могу придумать пример, когда клиенту (сервис bot) нужно знать id (из базы данных) ссылки, которую клиент получил. Я, конечно же, поправлю, но прошу помочь с пониманием этого момента.

edikgoose replied:
Если я правильно понял вопрос ("зачем возвращать id ссылки?"), то это нужно чтобы клиент твоего сервиса мог дальше работать с этой сущностью, так как этот id - это уникальный идентификатор ссылки в твоей системе, и его принимают все остальные методы (удаление, обновление)

edikgoose commented on scrapper/src/main/java/edu/java/api/LinksController.java:
```diff
@@ -19,45 +29,67 @@
 
     @PostMapping()
-    public ResponseEntity<?> handlePostLinks(@RequestHeader("Tg-Chat-Id") Long tgChatId) {
+    public ResponseEntity<?> addTrackingLinkForChat(
```
А почему здесь не LinkResponse?)

de-brasi replied:
Поправил, не помню почему так сделал, скорее всего просто недосмотрел)

edikgoose commented on scrapper/src/main/java/edu/java/api/LinksController.java:
```diff
@@ -19,45 +29,67 @@
     }
 
     @DeleteMapping()
-    public ResponseEntity<?> handleDeleteLinks(@RequestHeader("Tg-Chat-Id") Long tgChatId) {
```
И здесь тоже почему <?>?

de-brasi replied:
Исправил

edikgoose commented on scrapper/src/main/java/edu/java/clients/GitHubClient.java:
```diff
@@ -12,16 +12,14 @@
+    );
 
 
     public GitHubClient() {
```
У тебя три конструктора не используются, может их удалить лучше?

de-brasi replied:
В дз7 подчистил. Понимаю что не очень с точки зрения проверки так ссылаться на другие ветки, но опасаюсь что потом уж очень много конфликтов при слиянии решать надо будет, извини. Если настоятельно рекомендуешь исправить именно тут, то поправлю

edikgoose replied:
все ок, посмотрю в седьмой)

edikgoose commented on scrapper/src/main/java/edu/java/clients/GitHubClient.java:
```diff
@@ -76,6 +74,27 @@ public UpdateResponse fetchUpdate(String owner, String repo)
+    public UpdateResponse fetchUpdate(String url)
+        throws EmptyResponseBodyException, FieldNotFoundException {
+
+        Matcher matcher = RETRIEVE_GH_NAME_AND_REPO_NAME_FROM_URL.matcher(url);
```
У тебя получается, что клиент гитхаба уже отвечает за две вещи: отправить запрос на получение обновление и парсинг строки через регулярку. По хорошему это нужно разделить. Например, парсинг вынести в отдельный класс, который ответственен только за это.
Небольшой совет: над каждым классом коротко описывай что он делает. И в описании всегда должна быть одна ответственность. Если больше, то разделяй

edikgoose commented on scrapper/src/main/java/edu/java/clients/StackOverflowClient.java:
```diff
@@ -85,6 +84,22 @@ public UpdateResponse fetchUpdate(Integer questionId) throws EmptyResponseBodyEx
     }
 
+    public UpdateResponse fetchUpdate(String url) throws EmptyResponseBodyException, FieldNotFoundException {
+        Matcher matcher = RETRIEVE_QUESTION_NUMBER_FROM_URL.matcher(url);
```
То же самое, что и с гитхабом

edikgoose commented on scrapper/src/main/java/edu/java/configuration/BotClientConfig.java:
```diff
@@ -72,4 +74,34 @@ RestClient.ResponseSpec.ErrorHandler endpointUpdatesStatus5xxHandler(@Autowired
+
+        @Autowired
+        @Qualifier("endpointUpdatesStatus1xxHandler")
+        RestClient.ResponseSpec.ErrorHandler endpointUpdatesStatus1xxHandler,
```
Слушай, может это проще загнать вcю логику обработки в один ResponseErrorHandler, и его передавать в клиенте в методе onStatus. А то как-то громоздко выглядит

edikgoose commented on scrapper/src/main/java/edu/java/configuration/ClientConfiguration.java:
```diff
@@ -10,14 +10,14 @@
     @Bean("stackOverflowClient")
     public StackOverflowClient stackoverflowClient(
-        @Value("#{@stackoverflowClientSettings.timeInMilliseconds()}") int timeoutInMilliseconds
+        @Value("#{@stackoverflowClientSettings.timeoutInMilliseconds()}") int timeoutInMilliseconds
```
Удобнее будет, если ты создашь отдельный класс с анноташкой @ConfigurationProperties(prefix = "stackoverflowClientSettings") и в нем создать поле timeoutInMilliseconds. А то в твоем случае при добавлении новых полей придется везде писать  @Value("#{@stackoverflowClientSettings.name()}")

de-brasi replied:
Не до конца понял. Поправил, сделал два класса для ГХ и Стекорверфлоу с полем, геттером и сеттером, для получения параметром получаю класс с настройками как бин и через сеттеры беру значение. Правильно сделал, это имелось в виду?

edikgoose commented on scrapper/src/main/java/edu/java/domain/JdbcChatLinkBoundRepository.java:
```diff
@@ -0,0 +1,225 @@
+    @Transactional
+    public boolean add(ChatLinkBound chatLinkBound) {
+        /*
+        Add bound between chat and link in table 'track_info'.
```
Документацию лучше вынеси в интерфейс и используй javadoc. Ссылки на все исключения прописывай через {@link ...} и в целом соблюдай его формат

плюс дока пишется над методом, а не под ним как в питоне. В idea если ты напишешь /** над методом и нажмешь enter, она тебе сама нагенерирует

de-brasi replied:
Поправил, спасибо за совет)

edikgoose commented on scrapper/src/main/java/edu/java/domain/JdbcChatLinkBoundRepository.java:
```diff
@@ -0,0 +1,225 @@
+                linkRepository.add(chatLinkBound.link());
+            }
+
+            final String getTgChatIdQuery = "select id from telegram_chat where chat_id = ? limit 1";
```
Делаешь два одинаковых запроса в базу. Сначала считаешь кол-во и проверяешь, что такой чат есть, потом ищешь чат с таким id.
Можно один раз получить чат:
```sql
select id from telegram_chat where chat_id = ? limit 1
```
и проверить, что оно не null

edikgoose commented on scrapper/src/main/java/edu/java/domain/JdbcChatLinkBoundRepository.java:
```diff
@@ -0,0 +1,225 @@
+                chatLinkBound.chat().id()
+            );
+
+            final String getLinkIdQuery = "select id from links where url = ? limit 1";
```
То же самое

edikgoose commented on scrapper/src/main/java/edu/java/domain/JdbcChatLinkBoundRepository.java:
```diff
@@ -0,0 +1,225 @@
+            );
+
+            final String countSuchPairQuery =
+                "select count(*) from track_info where link_id = ? and telegram_chat_id = ?";
```
Лучше попробовать вставить элемент и если получить ошибку constraint'а, то отдать false
Лучше лишний раз не ходить в базу, если в этом нет необходимости. Условие, что link_id и telegram_chat_id уникальны у тебя прописаны на уровне базы, поэтому это надо использовать
UPD: ниже описано, как лучше

edikgoose commented on scrapper/src/main/java/edu/java/domain/JdbcChatLinkBoundRepository.java:
```diff
@@ -0,0 +1,225 @@
+
+    @Override
+    @Transactional
+    public @Nullable ChatLinkBound remove(ChatLinkBound chatLinkBound) {
```
Лучше используй Optional вместо Nullable

Можешь посмотреть контракты методов спрингового JpaRepository и оттуда взять

de-brasi replied:
Ниже есть совет возвращать число измененных записей в бд, поэтому для репозитория сделал как в совете ниже, так получилось удобнее вынести "сложную" логику в сервисы

edikgoose commented on scrapper/src/main/java/edu/java/domain/JdbcChatLinkBoundRepository.java:
```diff
@@ -0,0 +1,225 @@
+                chatLinkBound.chat().id(), chatLinkBound.link().uri().toURL().toString()
+            );
+
+            if (affectedRowsCount > 1) {
```
Я бы эту всю логику вынес в сервис, а методом возвращал кол-во измененнных сущностей.
Репозиторий должен просто предоставить удобное API для исполнения sql скриптов и ловить нужные исключения в случае ошибок с базой. А вся остальная логика с тем какие данные вернулись должна быть в сервисе

(так же и для всех остальных методов репозиториев)

de-brasi replied:
Поправил

edikgoose commented on scrapper/src/main/java/edu/java/domain/JdbcChatLinkBoundRepository.java:
```diff
@@ -0,0 +1,225 @@
+    public Collection<ChatLinkBound> search(Predicate<ChatLinkBound> condition) {
+        return findAll()
+            .stream()
+            .filter(condition)
```
Тут имелось в виду сделать фильтрацию на уровне запроса)
Для понимания, представь, что у тебя записей в базе на 2гб, а на хипу в JVM выделено только 512mb. Значит, что все записи из таблицы тупо не поместятся в памяти, поэтому их нужно отфильтровать на уровне базы и в память вытащить только нужные.

edikgoose commented on scrapper/src/main/java/edu/java/domain/JdbcChatLinkBoundRepository.java:
```diff
@@ -0,0 +1,225 @@
+
+    private final static Logger LOGGER = LogManager.getLogger();
+
+    private static class LinkRowMapper implements RowMapper<ChatLinkBound> {
```
Можно в бин вынести

edikgoose commented on scrapper/src/main/java/edu/java/domain/JdbcLinkRepository.java:
```diff
@@ -0,0 +1,129 @@
+    public boolean add(Link link) {
+
+        // TODO:
+        //  Почему-то если пытаться вставлять новую запись и ловить
```
Поскольку ты используешь @Transactional анноташку, все действия внутри метода происходят в транзакции и для всех них создается один коннекшн к базе.
При выполнении любого sql jdbcTemplate отлавливает ошибку sql, **закрывает коннекшн** (по факту роллбачит нашу транзакцию) и прокидывает ее выше. 

По поводу ошибок. По сути тебе надо сделать следующие вещи при добавлении ссылки:
- Проверить, что чат существует, иначе 404. (Простым SELECT'ом)
- Попробовать создать сущность ссылки. Тут поинтереснее, так как тут нужно поведение: создай, если отсутствует. Для этого в postgre есть [выражение](https://www.postgresql.org/docs/current/sql-insert.html#SQL-ON-CONFLICT):
```sql
INSERT ... ON CONFLICT DO NOTHING/UPDATE
```
которое буквально говорит: я попробую вставить, но если будут конфликты я ничего делать не буду (или обновлю что-то)
- Добавить новую связь. Тут можно по разному сделать, если есть такая: либо вернуть существующую, либо вернуть 400. Но поскольку ты используешь POST, а он не идемпотентный, то лучше наверное 400

de-brasi replied:
Спасибо за такое развернутое пояснение, когда с этой проблемой столкнулся (с/вы)горел пытаясь починить :), надобавлял "on conflict" везде где можно

edikgoose commented on scrapper/src/main/java/edu/java/domain/JdbcLinkRepository.java:
```diff
@@ -0,0 +1,129 @@
+        //      "
+        //  "На стороне БД" - будто бы имеется в виду что фильтрация
+        //  должна быть осуществлена самим постгресом, а для этого (при использовании jdbc, насколько я смог узнать)
+        //  нужно просто менять строку запроса; то есть как "предикат" - строка типа "where <some conditions>".
```
Вообще я это понял, как просто сделать 
```sql
SELECT * FROM link WHERE (currentTime - last_update_time) > ?
```

de-brasi replied:
Исправил

edikgoose commented on scrapper/src/main/java/edu/java/services/jdbc/JdbcLinkService.java:
```diff
@@ -0,0 +1,78 @@
+            boolean res = linkBoundRepository.add(bound);
+
+            if (!res) {
+                throw new ReAddingLinkException();
```
Как-то неочевидно сделано. А если в ответ false метода еще одна исключительная ситуация заложится. Лучше выкидывай исключения сразу в репозитории, а методе создания возвращай саму сущность

de-brasi replied:
Мне показалось, что эта рекомендация конфликтует с рек-ей выше "вынести логику в сервисы". Как сделал: репозиторий отдает число измененных записей, в сервисе проверяю это число, если оно != 1 то бросаю ошибку. Кажется так более явно, и в сервисе контролируется когда считать что произошла ошибка, а когда - ожидаемое поведение. Так нормально?

edikgoose commented on scrapper/src/main/java/edu/java/services/jdbc/JdbcLinkService.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    @Override
+    @Transactional
```
Если только читаешь, то лучше использовать @Transactional(readOnly = true)

edikgoose commented on scrapper/src/main/java/edu/java/services/jdbc/JdbcLinkUpdater.java:
```diff
@@ -0,0 +1,253 @@
+
+        Predicate<Link> outdatedLinkPredicate = link -> {
+            try {
+                return Duration.between(OffsetDateTime.now(), getLinkUpdateTime(link)).compareTo(updateInterval) > 0;
```
Вот это условие можно засунуть в сам sql запрос и сделать отдельным методом в репозитории

de-brasi replied:
Поправил, изначально интерпретировал задание не так как задумывалось автором. Сейчас добавил новый метод (JdbcLinkRepository::getOutdated) который строит запрос с ограничениями, учитывающими Duration в аргументе

edikgoose commented on scrapper/src/main/java/edu/java/domain/JdbcLinkRepository.java:
```diff
@@ -0,0 +1,129 @@
+        //  А так как написано сейчас не кажется удобным.
+        //  В любом случае - https://ru.wikipedia.org/wiki/%D0%AD%D1%82%D0%BE_%D0%BD%D0%B5%D0%BC%D0%BD%D0%BE%D0%B3%D0%BE,_%D0%BD%D0%BE_%D1%8D%D1%82%D0%BE_%D1%87%D0%B5%D1%81%D1%82%D0%BD%D0%B0%D1%8F_%D1%80%D0%B0%D0%B1%D0%BE%D1%82%D0%B0
+
+        return findAll()
```
На больших таблицах так вообще лучше никогда не делать. Да вообще лучше так не делать)

de-brasi replied:
Окей)

edikgoose commented on scrapper/src/main/java/edu/java/services/jdbc/JdbcLinkUpdater.java:
```diff
@@ -0,0 +1,253 @@
+            }
+        };
+
+        Collection<Link> toUpdate = getAllLinksFilteredByPredicate(outdatedLinkPredicate);
```
Назови уж лучше linksToUpdate)

edikgoose commented on scrapper/src/main/java/edu/java/services/jdbc/JdbcLinkUpdater.java:
```diff
@@ -0,0 +1,253 @@
+                boolean linkUpdated;
+                if (gitHubClient.checkURLSupportedByService(currentLinkUrl)) {
+                    final OffsetDateTime actualUpdateTime =
+                        gitHubClient.fetchUpdate(currentLinkUrl).updateTime();
```
Ходить в транзакции во внешние сервисы - RED FLAG. Запомни это. Из-за этого очень много бед происходит, так как транзакции становятся долгими -> долго держать коннекшн -> сильно уменьшается пропускная способность системы. 
Для более четкого понимания, представь, что у тебя 10 коннекшенов к базе. Прилетает одновременно 15 запросов на обновление ссылки гитхаба и 100 запросов на простое получение ссылок из базы. Гитхаб в этот день всем отвечает по 50 секунд и получается, что следующие 100 запросов будут ждать 50 секунд своего коннекшена
**Все походы во внешние сервисы - вне транзакции**

de-brasi replied:
Спасибо, поправил. Убрал @Transactional над update, делегировал обновление данных методам в JdbcLinkRepository, то есть походы в сторонние сервисы теперь вне транзакций. Тем не менее у меня остается работа с бд в методе getSubscribers, но кажется что это ок потому что запрос только на чтение и нет особенных причин заботиться о строгой консистентности полученных из БД данных.

edikgoose commented on scrapper/src/main/java/edu/java/LinkUpdaterScheduler.java:
```diff
@@ -1,15 +1,54 @@
-        System.out.println("Update");
+        LOGGER.info("Updating...");
+
+        Duration checkingDeadline = Duration.ofMinutes(1);
```
Лучше вынести в конфиг

de-brasi replied:
Поменял на значение forceCheckDelay у Scheduler бина в конфиге

edikgoose commented on scrapper/src/main/java/edu/java/services/jdbc/JdbcLinkUpdater.java:
```diff
@@ -0,0 +1,253 @@
+
+        Predicate<Link> outdatedLinkPredicate = link -> {
+            try {
+                return Duration.between(OffsetDateTime.now(), getLinkUpdateTime(link)).compareTo(updateInterval) > 0;
```
А почему ты проверяешь linkUpdateTime, а не lastCheckTime у ссылки? linkUpdateTime это дата последнего обновления на внешнем сервере, а lastCheckTime это когда ты последний раз проверял саму ссылку

de-brasi replied:
Просто ошибка из-за невнимательности) Исправил

edikgoose commented on scrapper/src/main/java/edu/java/services/jdbc/JdbcLinkUpdater.java:
```diff
@@ -0,0 +1,253 @@
+                    final OffsetDateTime actualUpdateTime =
+                        gitHubClient.fetchUpdate(currentLinkUrl).updateTime();
+                    linkUpdated = notifyClientsIfUpdatedTimeChanged(actualUpdateTime, link, subscribers);
+                } else if (stackOverflowClient.checkURLSupportedByService(currentLinkUrl)) {
```
Можно было заинжектить списко интерфейсов клиентов и по ним пройтись

de-brasi replied:
Я исправил это когда делал JPA сервис в Дз-6. Добавил еще один компонент, который как раз примерно так и работает. В этой ветке позволю себе оставить такой подход, все равно после в лива в коде будет актуальная версия

edikgoose reviewed (commented):
Прошу прощения за задержку в проверке. Вижу, что старался все сделать хорошо, но есть много замечаний:
- По ресту сущности должны быть в ед числе, поэтому поменяй пожалуйста все эндпоинты (например, POST /scrapper/api/links на POST /scrapper/api/link). И в целом во всех названиях должно быть ед число (таблицы, методы)
- Много тудушек сильно грязнят код, лучше заведи себе канбан где-нибудь и отслеживай все задачи там.
- Юзай sl4j анноташку вместо того, чтобы создавать логгер каждый раз
- Добавь, пожалуйста, id в hiber entity. Без него совсем не катит)
- Юзай @RequiredArgsContructor, чтобы не прописывать bolierplate конструкторы везде
- Метод поиска по критерию нужно переделать)

Снижу один балл за все замечания в сумме.

de-brasi committed:
remove debug log

de-brasi committed:
remove manually creating logger and mark classes via @Slf4j

de-brasi committed:
remove manually creating logger and mark classes via @Slf4j

de-brasi committed:
remove manually creating logger and mark classes via @Slf4j

de-brasi committed:
remove unnecessary logging

de-brasi committed:
better error message

de-brasi committed:
delete extra todo

de-brasi committed:
remove unnecessary logging

de-brasi committed:
remove unnecessary todos

de-brasi committed:
better naming (LinksController -> LinkController)

de-brasi commented:
> * По ресту сущности должны быть в ед числе, поэтому поменяй пожалуйста все эндпоинты (например, POST /scrapper/api/links на POST /scrapper/api/link). И в целом во всех названиях должно быть ед число (таблицы, методы)

Тоже этот момент смутил, но я решил сделать в точности как в OpenAPI спецификации в условии ДЗ-3 (https://gist.github.com/sanyarnd/52a78a01fa9ec234c8ad50fbc5ecc9e4). И, например, у SOF в апи для вопросов тоже эндпоинт в множественном числе: "https://api.stackexchange.com/2.3/questions/..."

de-brasi committed:
delete required args constructor and mark class with @RequiredArgsConstructor

de-brasi committed:
delete required args constructor and mark class with @RequiredArgsConstructor

de-brasi committed:
linter fix

de-brasi committed:
better naming

de-brasi committed:
add command start

de-brasi committed:
add transactional methods updateLastCheckTime and updateLastUpdateTime

de-brasi committed:
add tests for new methods updateLastCheckTime and updateLastUpdateTime

de-brasi committed:
make update function non-transactional, use transactional, methods updateLastCheckTime and updateLastUpdateTime (from JdbcLinkRepository) instead

de-brasi committed:
linter fix

de-brasi committed:
use check delay from properties

de-brasi committed:
sieve links by last check time instead of last update time

de-brasi committed:
add LinkResponse as parameter to generic ResponseEntity

de-brasi committed:
init

de-brasi committed:
fix time comparison in outdatedLinkPredicate

de-brasi committed:
move parsing to own class Parser

de-brasi committed:
move parsing to own class Parser

de-brasi committed:
linter fix

de-brasi committed:
linter fix

de-brasi committed:
fix parsed json key name

de-brasi committed:
check if response body is null

de-brasi committed:
remove unnecessary ObjectMapper

de-brasi committed:
merge all ErrorHandler beans into one ResponseErrorHandler;
use new BotClient constructor;

de-brasi committed:
use one ResponseErrorHandler instead of ErrorHandler

de-brasi committed:
linter fix (anonymous class to private record implements ResponseErrorHandler)

de-brasi committed:
linter fix

de-brasi committed:
init

de-brasi committed:
getting settings from property classes GitHubClientSettings and StackOverflowClientSettings instead of using @Value annotation

de-brasi committed:
extend EnableConfigurationProperties arguments with GitHubClientSettings.class and StackOverflowClientSettings.class

de-brasi committed:
better documentation

de-brasi committed:
fix method "add": avoid extra queries to database, add "on conflict do nothing" to sql for inserting pair that may be already existing

de-brasi committed:
move complex logic from repositories to services

de-brasi committed:
update test for simplified repositories logic

de-brasi committed:
remove returns Optional

de-brasi committed:
fix (repositories remove returns Optional)

de-brasi committed:
better

de-brasi committed:
change BaseEntityRepository interface and fix implementation classes:
- add returns int instead of boolean;
- remove returns int instead of Optional<Entity>

de-brasi committed:
fix tests

de-brasi committed:
add getOutdated method for filtering links in database by last_check_time field

de-brasi committed:
use LinkRepository's "getOutdated" method

de-brasi committed:
simplify methods: updateLastCheckTime, updateLastUpdateTime

de-brasi committed:
add check after updating last_update_time and last_check_time in method notifyClientsIfUpdatedTimeChanged

de-brasi committed:
remove checkChatExists method

de-brasi committed:
init

de-brasi committed:
add method Collection<String> getLinksTrackedBy(Long chatId);
use RowMappers as beans from JdbcRowMappersConfig

de-brasi committed:
use RowMappers as beans from JdbcRowMappersConfig

de-brasi committed:
add method getEntityId(E entity) for getting database's id of entity record

de-brasi committed:
add mapper to Link object

de-brasi committed:
add field id

de-brasi committed:
use RowMapper<Link> instead of RowMapper<String>;
fix method getLinksTrackedBy - now returns Collection<Link> (Collection<String> before)

de-brasi committed:
support new Link structure (id of link is taken into account);
better listAll method

de-brasi committed:
fix (reason is new Link structure)

de-brasi committed:
linter fix

de-brasi committed:
change id type from Long to Integer

de-brasi committed:
getEntityId returns Integer instead of Long

de-brasi committed:
fix Link id datatype from Long to Integer (for bean RowMapper<Link>)

de-brasi committed:
fix Link id datatype from Long to Integer

de-brasi committed:
in "update" method now working with links as Link class instant instead of String;
fix stub for id when send updates with BotClient

de-brasi committed:
delete unnecessary todo

de-brasi committed:
remove id stub for Links, use values from db
delete unnecessary todo

de-brasi committed:
remove extra imports

de-brasi committed:
typo
